### DISCUSSION TITLE
add DefaultNamespaceGeneration property in csproj

### DISCUSSION THREAD

**Discussion by burtonrodman** [dotnet/csharplang#4902] [Language Ideas] [answered]
The current behavior for a C# file with no namespace directive is to be placed in the `global` namespace.  I have never been in a shop where it is acceptable to place things in the global namespace.  Some analyzers even exist to identify and fix namespaces inconsistent with the project's name and folder structure (JetBrains??).

In order to reduce boilerplate and improve consistency of naming, would be great to be able to opt-in to default namespace generation.

PROPOSAL:
Add an option in csproj to specify a default namespace generation algorithm.  The default would remain global - you must opt-in to this feature.  One or more standard algorithms would be provided.  i.e. UseAssemblyName, UseAssemblyAndFolderName.

If nothing in the project is changed, the default would assumed to be:
`<DefaultNamespaceGeneration>UseGlobal</DefaultNamespaceGeneration>`

EXAMPLE:
MyProject.csproj
`<DefaultNamespaceGeneration>UseAssemblyAndFolderName</DefaultNamespaceGeneration>`

Controllers\UserController.cs
`public class UserController { }`

The resulting namespace would be `MyProject.Controllers.UserController`.

Bonus Points:

- allow adding new algorithms through the SourceGenerator infrastucture.
- Make it a compiler warning if a namespace is specified that does not match what would have been set by the specified algorithm.

**Comment by burtonrodman** [comment, marked as answer]
@bernd5 any comment as to why you downvoted?

**Reply by CyrusNajmabadi** [reply]
We prefer to avoid dialects of C# where the choices of options in your project file affect the end resultant metadata or IL you get.  This would generally go against this preference of ours.  There would have to be a super compelling reason why we'd go this direction here.  Cheers!

**Reply by burtonrodman** [reply]
I'm legitimately interested -- how is this any different than top-level programs?  it seems like a natural progression.

**Reply by CyrusNajmabadi** [reply]
Top level programs have no need for anything special in the project file.

**Reply by CyrusNajmabadi** [reply]
The code also does not live in a namespace. So it is at odds with the proposal here

**Reply by burtonrodman** [reply]
do you have any reference where I can learn about these design preferences?

**Reply by burtonrodman** [reply]
i emphatically see the namespace directive as boilerplate -- especially when it's become such a well accepted pattern that there's analyzers to enforce specific patterns.  do you have any suggestions that do not violate the design preferences?

**Reply by burtonrodman** [reply]
Source Generators require a reference in the csproj and also alter compiler output.  Does this not break your preference?

**Reply by HaloFour** [reply]
Tooling already handles this by automatically inserting the directory as the namespace in any project item added to the current project.  As such it's "boilerplate" that doesn't have to be written.  IMO this is not a problem that needs to be solved by the language or compiler as neither care about files or directory structures.

**Reply by burtonrodman** [reply]
generated code is still code that needs to read, understood and maintained -- this is a cost to the developer.  I understand (and agree) that the compiler may not be the right place for this.  I considered a Source generator, but as I understand Source Generators, they cannot alter existing source -- only add to it.  Otherwise, I would just write such a generator and move on.   Hence my question -- do you have any other suggestion?  And I would still love to have a reference to the team's design preferences.

**Reply by HaloFour** [reply]
> generated code is still code that needs to read, understood and maintained -- this is a cost to the developer.

We're talking a single line here, one that explicitly conveys a single thing, the namespace of the current code.  To have the compiler do anything else would be a further cost to the developer in understanding the options that exist and/or what implicit behaviors they bring.  IMO, it's not worth it, especially when the tooling generates that single line for you.  With C# 10 you don't even need the braces and indentation.

> do you have any other suggestion? 

Not really, no.

**Reply by HaloFour** [reply]
See also: https://github.com/dotnet/csharplang/issues/137#issuecomment-736761850

**Reply by burtonrodman** [reply]
no, we're talking 3 lines in billions of source files -- most of which are utterly unnecessary.  I was not aware of C# 10 File-level namespaces.  Thanks for your consideration.

and would STILL love a reference to the team's design preferences.

**Reply by HaloFour** [reply]
@CyrusNajmabadi is a member of the language team (not sure why he doesn't have a "Member" badge on his comments), as is @333fred which I referenced.  As the latter states the compiler itself doesn't follow the a strict directory structure to namespace convention.  Neither do many of the projects I've worked on.

**Reply by burtonrodman** [reply]
thanks.  every project I've been on recently has strived to follow this pattern -- i realize the compiler might be a slightly more complex animal ;)

I might not be expressing myself well wrt a reference to the team's design preferences -- can you share a link to the documentation describing ALL of the team's general language design preferences -- for example, @CyrusNajmabadi said "We prefer to avoid dialects of C#...".  Where is this rule and its rationale expressed in publicly available documentation?

**Reply by yaakov-h** [reply]
I find that lot of projects follow this pattern because of influences deriving from Java, not necessarily because its makes sense for the particular namespaces and folder structure in the project itself.

**Reply by HaloFour** [reply]
> "We prefer to avoid dialects of C#...". Where is this rule and its rationale expressed in publicly available documentation?

I'm not aware of such a document.  For those of us that follow the repo we understand the team's reluctance and high bar to accept compiler options that change the behavior of the language just from following the conversations here.  Every such option doubles the flavors of the language and makes it much more difficult to reason about code without knowing how the project is setup.

**Reply by burtonrodman** [reply]
@yaakov-h it is a common pattern none the less.  and it does make sense in many projects -- especially wrt to projects that follow patterns like MVC.

**Reply by burtonrodman** [reply]
@HaloFour could such a document be created so that occasional C# proposal contributors like myself can make more educated proposals in the future?

**Reply by burtonrodman** [reply]
@HaloFour I feel that if I were not one to be forward about requesting a reasoning why this got 5 downvotes, I would have gone away dejected (as I am anyway), but never bothered to participate in this forum again.  I would suggest that team members not downvote without an explanation in the future to maintain a psychologically safe and open community.

**Reply by burtonrodman** [reply]
@ZacharyPattern care to explain?

**Reply by CyrusNajmabadi** [reply]
I don't think there's a problem with upvotes/downvotes.  It's a generally reasonable pattern for people to express their general support (or lack thereof) for a particular idea.  It can help get an initial gauge on how many others in the community feel about things by default.

Don't be dejected by lots of downvotes though.  I'm on the team and i've had far more rejected than accepted proposals myself :)

**Reply by CyrusNajmabadi** [reply]
Note: you might find it more conducive initially to discuss things over at gitter.im/dotnet/csharplang or discord.gg/csharp.  Those are two communities where you'll find a lot more active discussion around things :)

**Reply by burtonrodman** [reply]
I fully expect to be required to defend my proposal and to have some opposition to a proposal of this nature.  What I don't appreciate is a downvote with no explanation -- that's not helpful to anyone.

**Reply by CyrusNajmabadi** [reply]
> - that's not helpful to anyone.

I think it helps serve as just a general 'read of hte room' as it were :)  

Don't worry though, people will definitely let you know (as i did) the reasons for why they do or don't support.  But we foster a community here where it's ok for people to include lots of information, or just weigh in with a general yes/no on how they feel about things.  Thanks!

If you'd like any more info, i'd be happy to share it!

**Reply by burtonrodman** [reply]
but that's exactly my point -- would you have let me know if i wasn't one to be forward about requesting the reason?  or would this whole conversation have died on the vine?  _as team members_, I would just suggest that simply downvoting as the initial reaction is not sufficient.

**Reply by 333fred** [reply]
Only one person who is on the team has voted, and Cyrus has given you his feedback. No one else in this conversation (besides me) is from Microsoft or is a member of the LDM.

**Reply by ZacharyPatten** [reply]
> could such a document be created so that occasional C# proposal contributors like myself can make more educated proposals in the future?

You are looking at it. That is what these discussions are for. If you do a search for "dialect" on this repo you will find multiple discussions where the same sentiment is shared: https://github.com/dotnet/csharplang/search?q=dialect&type=discussions. Should an "official" document be created? No, because opinions can changed. Just look at nullable reference types. Although the community is currently generally opposed to multiple dialects in C#, an exception was made for nullable reference types because the positives vastly outweighed the negatives for that feature/topic.

>  I would suggest that team members not downvote without an explanation in the future...

The people who maintain Roslyn are extremely busy. Now that the code is open source and on GitHub, they are engaging with the community more than ever _(thanks for all the hard work guys)_. :) If the Roslyn developers had to take the time to write a paper for every GitHub discussion, we would never get any new features. Cut them some slack. Don't take down votes personally. They are a downvote of the idea, not of the individual that is proposing them. Not only is there GitHub, but many of the Roslyn developers are active on other platforms such as Discord. So it is not like there is limited engagement at the moment.

As for your topic for this discussion, is there a benefit under the right circumstances? Sure. However, there just isn't enough benefit to justify it currently, and it would probably do more harm than good in my opinion.

**Reply by CyrusNajmabadi** [reply]
I'm the only team member here.  :)

> I would just suggest that simply downvoting 

FWIW, i didn't just downvote.  I gave my explanations for why i did, along with a willingness to discuss this at whatever level to help with understanding and furthering this :)

That said, i think it's acceptable for people to just give an initial 'read of the room'-style view to give a quick assessment on how people think about this.  I'm not comfortable telling people they can't do this or that they need to be forced to add more than that.  i think that might be stifling and that people should be free to give this sort of view.  Trust me when i say that htere will certainly be enough people willing to discuss the issue in depth to help with developing the discussion. 

Thanks!

**Reply by CyrusNajmabadi** [reply]
>  Not only is there GitHub, but many of the Roslyn developers are active on other platforms such as Discord. So it is not like there is limited engagement at the moment.

As mentioned in https://github.com/dotnet/csharplang/discussions/4902#discussioncomment-967982 I recommend gitter.im/dotnet/csharplang and discord.gg/csharp

**Reply by RikkiGibson** [reply]
I can sympathize that receiving several downvotes on your proposal can be discouraging. I encourage you to not take it personally.

I would just say that coming up with successful proposals for the language is **hard**. It requires picking up a lot of background about the LDM's preferences and priorities, as well as successfully competing for the language design team's time, who by necessity have to turn down or push out for years features that could be helpful to someone but just don't meet the cost/benefit threshold for inclusion.

 I don't think it's practical to fully document the language team's preferences, unfortunately, although if someone were to link to some "design principles for C#" type documents from current or past language designers, I'm sure that would be helpful. It would probably also be helpful to look through some of the past accepted or rejected proposals as well as the LDM notes about those proposals to see why things were done the way they were or why things were rejected.

Regarding this feature specifically: I think that an analyzer that enforces some namespace name conventions is probably adequate to fulfill the goals of the feature. I don't think that there's a concern with, for example, each file needing to contain a namespace declaration which matches the project name + folder path. It feels like for each person that sees the namespace declaration as noise code, another person is likely to say: great, the fully qualified names of the types in this file are apparent from the file itself. As mentioned above, the C# 10 "file scoped namespaces" should also help a great deal with reducing the noise.

**Comment by hawkerm** [comment]
If it helps, I was just looking for something like this. Writing a simple library and just want all the classes to be in a common shared namespace. Would be nice to have it in one spot in the project file vs. in every source file.

Seems like the inverse and an extension of how global usings can be defined in project files now as well? [See Using Item in msbuild docs here](https://learn.microsoft.com/dotnet/core/project-sdk/msbuild-props#using).
